### ISSUE TITLE
Custom eval metric warnings and not using all cores

### ISSUE THREAD

**Issue by swapnil96** [catboost/catboost#2109]
Problem: Custom eval metric giving the below warning and also maybe for the same issue not using all the cores
catboost version: 1.0.6 and also tried 1.0.5
Operating System: Linux 3.10.0-862.3.3.el7.x86_64
CPU: Intel(R) Xeon(R) CPU E5-2665 0 @ 2.40GHz x86_64
GPU: None

The issue was not there like 10 days ago, also logs point to expit, but its not the issue as I tried removing that. Basically any code there produces the issue if it accesses the input params.
Some issue with params maybe, not sure

My custom metric is

`

    class CustomLoss2Class(object):
    
        def is_max_optimal(self):
            return False

        def evaluate(self, approxes, target, weight):
            y_true = target.astype(int)
            y_pred = np.rint(expit(approxes[0]))
            cm = confusion_matrix(y_true, y_pred)
            fnr = cm[1,0]/np.sum(cm[1,:])
            fpr = cm[0,1]/np.sum(cm[0,:])
            return 7*fnr + 1.2*fpr + 1.3*abs(fnr-fpr), 1
            
        def get_final_error(self, error, weight):
            return error
`

/home/sdas/miniconda3/lib/python3.9/site-packages/catboost/core.py:1619: UserWarning: Failed to optimize method "evaluate" in the passed object:
Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'expit': Cannot determine Numba type of <class 'numpy.ufunc'>

        y_true = target.astype(int)
        y_pred = np.rint(expit(approxes[0]))
        ^

  self._object._train(train_pool, test_pool, params, allow_clear_pool, init_model._object if init_model else None)

**Comment by ek-ak** [comment]
Hello! CatBoost uses Numba to jit custom user's  metrics.  Numba doesn't work well with scikit-learn.  You can rewrite you metric without sklearn (not sure that this will accelerate training) or specify metric-period to not calculate metric on each iteration.

**Comment by swapnil96** [comment]
But my issue is not related to sklearn I think because I have used numpy version of expit (simple sigmoid calculation), same warning persist then too. Its like you do any calculation under the evaluate fn, it will give this warning and it was not there like 20 days ago

**Comment by swapnil96** [comment]
Ex: 



    class CustomLoss2Class(object):

        def is_max_optimal(self):
            return False

        def evaluate(self, approxes, target, weight):
            y_true = target.astype(int)
            # y_pred = np.rint(expit(approxes[0]))
            # cm = confusion_matrix(y_true, y_pred)
            # fnr = cm[1,0]/np.sum(cm[1,:])
            # fpr = cm[0,1]/np.sum(cm[0,:])
            # return 7*fnr + 1.2*fpr + 1.3*abs(fnr-fpr), 1
            return 1, 1
        
        def get_final_error(self, error, weight):
            return error

Now I have commented all the logic, just reading the params and returning 1, 1, still the warning is there but now with a different message

```
/home/sdas/miniconda3/lib/python3.9/site-packages/catboost/core.py:1705: UserWarning: Failed to optimize method "evaluate" in the passed object:
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of BoundFunction(array.astype for array(float32, 1d, C)) with parameters (Function(<class 'int'>))

During: resolving callee type: BoundFunction(array.astype for array(float32, 1d, C))
During: typing of call at /tmp/ipykernel_12644/1280109660.py (7)


File "../../../tmp/ipykernel_12644/1280109660.py", line 7:
<source missing, REPL/exec in use?>

  self._object._train(train_pool, test_pool, params, allow_clear_pool, init_model._object if init_model else None)
```

**Comment by Joe-Khawand** [comment]
Has this been solved ?
